### PR TITLE
jit: support for memory grow/size

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: wasm/jit/jit_amd64.go [hunk 1]
```diff
@@ -1315,7 +1339,7 @@ func (b *amd64Builder) allocateRegister(t generalPurposeRegisterType) (reg int16
 	return
 }
 
-func (b *amd64Builder) setJITStatus(status jitCallStatusCode) *obj.Prog {
```
Comment (mathetake) [inline wasm/jit/jit_amd64.go:1318]: note: while here, I'm removing unused returns

File: wasm/jit/engine.go [hunk 2]
```diff
@@ -420,17 +422,23 @@ func (e *engine) exec(f *compiledWasmFunction) {
...
 	}
 	// If exceeds the max of memory size, we push -1 according to the spec
-	if uint64(newPages*wasm.PageSize+uint64(len(m.Buffer))) > max {
+	if uint64(newPages*wasm.PageSize+uint64(len(f.memory.Buffer))) > max {
 		v := int32(-1)
 		e.push(uint64(v))
 	} else {
-		e.push(uint64(uint64(len(m.Buffer)) / wasm.PageSize))
-		m.Buffer = append(m.Buffer, make([]byte, newPages*wasm.PageSize)...)
+		e.push(uint64(uint64(len(f.memory.Buffer)) / wasm.PageSize))
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go:436]: nit: re-use `f.builtinFunctionMemorySize() // grow returns the prior memory size on change`

File: wasm/jit/engine.go [hunk 3]
```diff
@@ -420,17 +422,23 @@ func (e *engine) exec(f *compiledWasmFunction) {
...
+	newPages := e.pop()
 	max := uint64(math.MaxUint32)
-	if m.Max != nil {
-		max = uint64(*m.Max) * wasm.PageSize
+	if f.memory.Max != nil {
+		max = uint64(*f.memory.Max) * wasm.PageSize
 	}
 	// If exceeds the max of memory size, we push -1 according to the spec
-	if uint64(newPages*wasm.PageSize+uint64(len(m.Buffer))) > max {
+	if uint64(newPages*wasm.PageSize+uint64(len(f.memory.Buffer))) > max {
```
Comment (codefromthecrypt) [inline wasm/jit/engine.go:432]: could newPages be zero? what should we do if that were the case?

File: wasm/jit/engine_test.go [hunk 4]
```diff
@@ -63,6 +63,34 @@ wasm backtrace:
...
+	// Then grow the memory.
+	const newPages uint64 = 10
+	out, _, err = store.CallFunction("test", "grow", newPages)
+	require.NoError(t, err)
+	// Grow returns the previous number of memory pages, namely zero.
+	require.Equal(t, uint64(0), out[0])
+	// Now size should return the new pages -- 10.
+	out, _, err = store.CallFunction("test", "size")
+	require.NoError(t, err)
+	require.Equal(t, newPages, out[0])
```
Comment (codefromthecrypt) [inline wasm/jit/engine_test.go:91]: I suppose there's no test we can do to see that the memory is usable? (ex using an offset that would only work if it was 10)

File: wasm/jit/jit_amd64.go [hunk 5]
```diff
@@ -1096,6 +1096,30 @@ func (b *amd64Builder) handleLe(o *wazeroir.OperationLe) error {
...
+	b.setJITStatus(jitCallStatusCodeCallBuiltInFunction)
+	// Set the function index.
+	b.setFunctionCallIndexFromConst(index)
+	// Release all the registers as our calling convention requires the callee-save.
+	b.releaseAllRegistersToStack()
+	// Set the continuation offset on the next instruction.
+	b.setContinuationOffsetAtNextInstructionAndReturn()
+	// Once we return from the function call,
+	// we must setup the reserved registers again.
+	b.initializeReservedRegisters()
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1120]: I like the intent to comment, but the only meaningful ones that say something more than what the function names say are.. 

```
// Release all the registers as our calling convention requires the callee-save.
```
and

```
Once we return from the function call,
 	// we must setup the reserved registers again.
```

can you remove the redundant ones as all lines are debt even comments unless they perform some good :D

File: wasm/jit/jit_amd64_test.go [hunk 6]
```diff
@@ -1404,6 +1402,58 @@ func TestAmd64Builder_handleCall(t *testing.T) {
...
 }
 
+func TestAmd64Builder_handleMemoryGrow(t *testing.T) {
+	builder := requireNewBuilder(t)
+
+	builder.initializeReservedRegisters()
+	// Emit memory.grow instructions.
+	builder.handleMemoryGrow()
+
+	// Compile.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:1412]: I see this often.. a comment compile for a function assemble.. should we rename assemble to compile?

File: wasm/jit/jit_amd64_test.go [hunk 7]
```diff
@@ -1404,6 +1402,58 @@ func TestAmd64Builder_handleCall(t *testing.T) {
...
+
+	// Run code.
+	eng := newEngine()
+	mem := newMemoryInst()
+	jitcall(
+		uintptr(unsafe.Pointer(&code[0])),
+		uintptr(unsafe.Pointer(eng)),
+		uintptr(unsafe.Pointer(&mem.Buffer[0])),
+	)
+	// The function call must end with jitCallStatusCodeCallBuiltInFunction status.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:1424]: change this to answer why or remove it as the assertion already says what

File: wasm/jit/jit_amd64_test.go [hunk 8]
```diff
@@ -1404,6 +1402,58 @@ func TestAmd64Builder_handleCall(t *testing.T) {
...
+	// Plus the index of builtin function must be that of memoryGrow.
+	require.Equal(t, int64(builtinFunctionIndexMemoryGrow), eng.functionCallIndex)
+}
+
+func TestAmd64Builder_handleMemorySize(t *testing.T) {
+	builder := requireNewBuilder(t)
+	builder.initializeReservedRegisters()
+	// Emit memory.size instructions.
+	builder.handleMemorySize()
+	// At this point, the size of memory should be pushed onto the stack.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:1435]: ex this is a good "what" comment because the assertion is a bit indirect. so this one helps a lot

**Review by codefromthecrypt — APPROVED** [review]
looks good, just a few notes to consider

File: wasm/jit/jit_amd64.go (hunk 5 above)
Comment (mathetake) [reply wasm/jit/jit_amd64.go]: 👍  https://github.com/tetratelabs/wazero/pull/85/commits/1b6114f43181c5f4c39e34e54a9c0cb86c90ab4f

File: wasm/jit/engine.go (hunk 2 above)
Comment (mathetake) [reply wasm/jit/engine.go]: https://github.com/tetratelabs/wazero/pull/85/commits/1b6114f43181c5f4c39e34e54a9c0cb86c90ab4f

File: wasm/jit/engine.go (hunk 3 above)
Comment (mathetake) [reply wasm/jit/engine.go:432]: zero is valid and even in that case this function works as expected. Added the test case: https://github.com/tetratelabs/wazero/pull/85/commits/19f32ca4cdc9d55c4219d0590fbd5ca68ef608c8

File: wasm/jit/jit_amd64_test.go (hunk 6 above)
Comment (mathetake) [reply wasm/jit/jit_amd64_test.go:1412]: https://github.com/tetratelabs/wazero/pull/85/commits/4449034a82cb32489f3aec84601ae80b6029f9ae agreed so renamed! :D

File: wasm/jit/jit_amd64_test.go (hunk 8 above)
Comment (mathetake) [reply wasm/jit/jit_amd64_test.go:1435]: agreed yeah:D I've put too many comments including the obvious ones. Will keep each line more meaningful from now on!

File: wasm/jit/jit_amd64_test.go (hunk 7 above)
Comment (mathetake) [reply wasm/jit/jit_amd64_test.go]: deleted https://github.com/tetratelabs/wazero/pull/85/commits/8389a7f67a6fd22b32c71996940f2108d7373f46

File: wasm/jit/engine_test.go (hunk 4 above)
Comment (mathetake) [reply wasm/jit/engine_test.go:91]: indirectly tested as the size instruction takes len of the memory region, meaning that these offsets can be assumed to work. (Plus memory load,store instructions are not implemented so I cannot write in-Wasm memory access to verify it for now :D)

**Comment by mathetake** [comment]
@codefromthecrypt thanks a lot! 🕺

**Pull request merged**